target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts

```diff
@@ -0,0 +1,361 @@
+	};
+
+	sfp_i2c: sfp-i2c {
+		compatible = "i2c-gpio";
+		sda-gpios = <&cp0_gpio1 0 (GPIO_ACTIVE_HIGH|GPIO_OPEN_DRAIN)>;
+		scl-gpios = <&cp0_gpio1 1 (GPIO_ACTIVE_HIGH|GPIO_OPEN_DRAIN)>;
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&cp0_gpio1 28 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_user: user {
+			label = "green:user";
+			gpios = <&cp0_gpio2 26 GPIO_ACTIVE_LOW>;
+		};
+
+		sfp {
+			label = "green:sfp";
+			gpios = <&cp0_gpio2 25 GPIO_ACTIVE_LOW>;
+		};
+
+		hdr1 {
+			label = "blue:hdr1";
+			gpios = <&cp0_gpio1 4 GPIO_ACTIVE_LOW>;
+		};
+
+		hdr2 {
+			label = "blue:hdr2";
+			gpios = <&cp0_gpio2 19 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	sfp: sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&sfp_i2c>;
+		mod-def0-gpios = <&cp0_gpio1 11 GPIO_ACTIVE_LOW>;
+		los-gpios = <&cp0_gpio1 2 GPIO_ACTIVE_HIGH>;
+		tx-fault-gpios = <&cp0_gpio1 6 GPIO_ACTIVE_HIGH>;
+		tx-disable-gpios = <&cp0_gpio1 5 GPIO_ACTIVE_HIGH>;
+		rate-select0-gpios = <&cp0_gpio1 3 GPIO_ACTIVE_HIGH>;
+	};
+};
+
+&uart0 {
+	status = "okay";
+
+	pinctrl-0 = <&uart0_pins>;
+	pinctrl-names = "default";
+};
+
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <20000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				compatible = "mikrotik,routerboot-partitions";
+				#address-cells = <1>;
+				#size-cells = <1>;
+				label = "MikroTik";
+				reg = <0x0 0xfe0000>;
+
+				hard_config: hard_config {
+					read-only;
+
+					nvmem-layout {
+						compatible = "mikrotik,routerboot-nvmem";
+
+						macaddr_hard: base-mac-address {
+							#nvmem-cell-cells = <1>;
+						};
+					};
+				};
+
+				soft_config {
+				};
+
+				dtb_config {
+					read-only;
+				};
+			};
+
+			partition@b0000 {
+				label = "RouterBOOT-primary";
+				reg = <0xb0000 0x10000>;
+			};
+
+			/* Empty space on NOR repurposed for U-Boot environment */
+			partition@fe0000 {
+				compatible = "u-boot,env";
+				label = "u-boot-env";
+				reg = <0xfe0000 0x20000>;
+			};
+		};
+	};
+};
+
+&cp0_nand_controller {
+	status = "okay";
+
+	nand@0 {
+		reg = <0>;
+		nand-rb = <0>;
+		nand-ecc-mode = "hw";
+		nand-ecc-strength = <4>;
+		nand-ecc-step-size = <512>;
+		nand-on-flash-bbt;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "YAFFS";
+				reg = <0x0 0x800000>;
+			};
+
+			partition@800000 {
+				label = "ubi";
+				reg = <0x800000 0x3f800000>;
+			};
+		};
+	};
+};
+
+&cp0_gpio2 {
+	enable-usb-power {
+		gpio-hog;
+		gpios = <23 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "enable USB power";
+	};
+
+	enable-leds-power {
+		gpio-hog;
+		gpios = <27 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "enable LED-s power";
+	};
```

Comment:
Hogging power switches has always annoyed me.
I might see if I can get [`regulator-output`](https://elixir.bootlin.com/linux/latest/B/ident/regulator-output) to work for these?
I looked at (USB power switch regulator) [some time ago for a different device](https://github.com/openwrt/openwrt/issues/10904#issuecomment-1312645744)

Would also like to see the device-specific gpios named in `gpio-line-names`. Will look at that too.

Reply:
userspace regulator works for me, allows control, and uses upstream driver.
guessed LED voltage here. Not concerned if you don't use it, or don't need attribution if you do.

```patch
commit 60205e998580dd23d8bf58da058d2e952a52fd26
Author: John Thomson <git@johnthomson.fastmail.com.au>
Date:   Sun Jun 23 08:42:59 2024 +1000

    mvebu: rb5009: userspace regulator for power switches
    
    Use userspace controlled regulators to manage USB port power and LED
    power switches.
    
    echo "disabled">/sys/bus/platform/drivers/reg-userspace-consumer/output-usb-port-power/state
    sleep 1
    echo "enabled">/sys/bus/platform/drivers/reg-userspace-consumer/output-usb-port-power/state
    echo "disabled">/sys/bus/platform/drivers/reg-userspace-consumer/output-led-power/state
    
    Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

diff --git a/target/linux/mvebu/cortexa72/config-6.6 b/target/linux/mvebu/cortexa72/config-6.6
index de2cec6f26ba..19ca2b29d1b1 100644
--- a/target/linux/mvebu/cortexa72/config-6.6
+++ b/target/linux/mvebu/cortexa72/config-6.6
@@ -102,6 +102,7 @@ CONFIG_QUEUED_SPINLOCKS=y
 CONFIG_RAS=y
 # CONFIG_RAVE_SP_CORE is not set
 CONFIG_REGULATOR_GPIO=y
+CONFIG_REGULATOR_USERSPACE_CONSUMER=y
 # CONFIG_RODATA_FULL_DEFAULT_ENABLED is not set
 CONFIG_SENSORS_IEI_WT61P803_PUZZLE_HWMON=y
 CONFIG_SERIAL_DEV_BUS=y
diff --git a/target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts b/target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts
index 886e87222b3a..9cc5ce22b98c 100644
--- a/target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts
+++ b/target/linux/mvebu/files-6.6/arch/arm64/boot/dts/marvell/armada-7040-rb5009.dts
@@ -76,6 +76,36 @@
 		tx-disable-gpios = <&cp0_gpio1 5 GPIO_ACTIVE_HIGH>;
 		rate-select0-gpios = <&cp0_gpio1 3 GPIO_ACTIVE_HIGH>;
 	};
+
+	reg_usb_port: regulator-usb {
+		compatible = "regulator-fixed";
+		regulator-name = "USB-power";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpio = <&cp0_gpio2 23 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-boot-on;
+	};
+
+	output-usb-port-power {
+		compatible = "regulator-output";
+		vout-supply = <&reg_usb_port>;
+	};
+
+	reg_leds: regulator-leds {
+		compatible = "regulator-fixed";
+		regulator-name = "LED-power";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&cp0_gpio2 27 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-boot-on;
+	};
+
+	output-led-power {
+		compatible = "regulator-output";
+		vout-supply = <&reg_leds>;
+	};
 };
 
 &uart0 {
@@ -177,22 +207,6 @@
 	};
 };
 
-&cp0_gpio2 {
-	enable-usb-power {
-		gpio-hog;
-		gpios = <23 GPIO_ACTIVE_HIGH>;
-		output-high;
-		line-name = "enable USB power";
-	};
-
-	enable-leds-power {
-		gpio-hog;
-		gpios = <27 GPIO_ACTIVE_HIGH>;
-		output-high;
-		line-name = "enable LED-s power";
-	};
-};
-
 &cp0_usb3_1 {
 	status = "okay";
 };
```

Reply:
Thanks, I wired up the USB regulator as the supply to the USB connector so the host controller driver will enable it automatically.

For LED-s I used your userspace regulator method.

While I was here, noticed that SERDES lanes were not described in Linux nor UTMI PHY was enabled so sorted that out.